Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx

```diff
@@ -230,14 +266,288 @@ ContourExtractor2DImageFilter<TInputImage>::GenerateData()
         break;
     } // switch squareCase
     progress.CompletedPixel();
-  } // iteration
+  } // pixel square iteration
 
   // Now create the outputs paths from the deques we've been using.
   this->FillOutputs();
-  m_Contours.clear();
-  m_ContourStarts.clear();
-  m_ContourEnds.clear();
-  m_NumberOfContoursCreated = 0;
+}
+
+
+template <typename TInputImage>
+void
+ContourExtractor2DImageFilter<TInputImage>::GenerateDataForLabels()
+{
+  using IndexType = typename InputRegionType::IndexType;
+  using RadiusType = typename ConstShapedNeighborhoodIterator<InputImageType>::RadiusType;
+  using RegionConstIterator = ImageRegionConstIterator<InputImageType>;
+  using ConstImageRegionRange = ImageRegionRange<const InputImageType>;
+  using ImageRegionRange = ImageRegionRange<InputImageType>;
+  using SizeType = typename InputRegionType::SizeType;
+  using SquareIterator = ConstShapedNeighborhoodIterator<InputImageType>;
```

Comment:
It might be interesting to use the new (ITK5) `itk::ShapedImageNeighborhoodRange` (possibly in combination with `itk::IndexRange`) in this case, instead of `ConstShapedNeighborhoodIterator`. I expect the retrieval of pixel values with the new range types to be significantly faster than the current `it.GetPixel(i)` calls in your pull request.

Reply:
This is now Issue #2348.